src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+        }
+
+        /// <inheritdoc/>
+        public void Add(IEdmChangedObject<TStructuralType> item)
+        {
+            InsertItem(_items.Count, item);
+        }
+       
+        /// <inheritdoc/>
+        public bool Contains(IEdmChangedObject<TStructuralType> item)
+        {
+            return _items.Contains(item);
+        }
+
+        /// <inheritdoc/>
+        public void CopyTo(IEdmChangedObject<TStructuralType>[] array, int arrayIndex)
+        {
+            _items.CopyTo(array, arrayIndex);
+        }
+
+        /// <inheritdoc/>
+        public bool IsReadOnly
+        {
+            get { return false; }
+        }
+
+        /// <inheritdoc/>
+        public bool Remove(IEdmChangedObject<TStructuralType> item)
+        {
+            int count = _items.Count;
+            RemoveItem(_items.IndexOf(item));
+
+            return count == _items.Count + 1;
+        }
+
+        /// <inheritdoc/>
+        IEnumerator<IEdmChangedObject<TStructuralType>> IEnumerable<IEdmChangedObject<TStructuralType>>.GetEnumerator()
+        {
+            return _items.GetEnumerator();
+        }
+
+        /// <inheritdoc/>
+        protected override void InsertItem(int index, IEdmChangedObject item)
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            if (_item == null)
+            {
+                throw Error.Argument("item", SRResources.ChangedObjectTypeMismatch, typeof(TStructuralType), item.GetType());
+            }
+
+            _items.Add(_item);
+            base.InsertItem(index, item);
+        }
+
+        /// <inheritdoc/>
+        protected override void ClearItems()
+        {
+            _items.Clear();
+        }
+
+        /// <inheritdoc/>
+        protected override void RemoveItem(int index)
+        {
+            Contract.Assert(_items.Count > index);
+            
+            _items.RemoveAt(index);
+            base.RemoveItem(index);
+        }
+
+        /// <inheritdoc/>
+        protected override void SetItem(int index, IEdmChangedObject item)
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            Contract.Assert(_item != null);
+
+            _items[index] = _item;
+        }
+
+        /// <summary>
+        /// Copy changed values is an implementation of Patch
+        /// </summary>
+        /// <param name="original">Original collection of the Type which needs to be updated</param>              
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        internal EdmChangedObjectCollection<TStructuralType> CopyChangedValues(ICollection<TStructuralType> original)
+        {
+            //Here we need to Find the key of the Type, then only we will be able to find from the collection that which item in the collection
+            //corresponds to the item in delta list(Edmchangedobjectcoll).
```

Comment:
Nit: wording is awkward. I think what you mean is "First find the key of the type so that we can find the corresponding item in the delta list."